src/Tools/dotnet-user-jwts/src/Helpers/DevJwtCliHelpers.cs

```diff
@@ -0,0 +1,166 @@
+    {
+        // Create signing material and save to user secrets
+        var newKeyMaterial = System.Security.Cryptography.RandomNumberGenerator.GetBytes(DevJwtsDefaults.SigningKeyLength);
+        var secretsFilePath = PathHelper.GetSecretsPathFromSecretsId(userSecretsId);
+
+        IDictionary<string, string> secrets = null;
+        if (File.Exists(secretsFilePath))
+        {
+            using var secretsFileStream = new FileStream(secretsFilePath, FileMode.Open, FileAccess.Read);
+            if (secretsFileStream.Length > 0)
+            {
+                secrets = JsonSerializer.Deserialize<IDictionary<string, string>>(secretsFileStream) ?? new Dictionary<string, string>();
+            }
+        }
+
+        secrets ??= new Dictionary<string, string>();
+
+        if (reset && secrets.ContainsKey(DevJwtsDefaults.SigningKeyConfigurationKey))
+        {
+            secrets.Remove(DevJwtsDefaults.SigningKeyConfigurationKey);
+        }
+        secrets.Add(DevJwtsDefaults.SigningKeyConfigurationKey, Convert.ToBase64String(newKeyMaterial));
+
+        using var secretsWriteStream = new FileStream(secretsFilePath, FileMode.OpenOrCreate, FileAccess.ReadWrite);
+        JsonSerializer.Serialize(secretsWriteStream, secrets);
+
+        return newKeyMaterial;
+    }
+
+    public static string GetApplicationUrl(string project)
+    {
+        ArgumentException.ThrowIfNullOrEmpty(nameof(project));
+
+        var launchSettingsFilePath = Path.Combine(Path.GetDirectoryName(project)!, "Properties", "launchSettings.json");
+        if (File.Exists(launchSettingsFilePath))
+        {
+            using var launchSettingsFileStream = new FileStream(launchSettingsFilePath, FileMode.Open, FileAccess.Read);
+            if (launchSettingsFileStream.Length > 0)
+            {
+                var launchSettingsJson = JsonDocument.Parse(launchSettingsFileStream);
+                if (launchSettingsJson.RootElement.TryGetProperty("profiles", out var profiles))
+                {
+                    var profilesEnumerator = profiles.EnumerateObject();
+                    foreach (var profile in profilesEnumerator)
+                    {
+                        if (profile.Value.TryGetProperty("commandName", out var commandName))
+                        {
+                            if (commandName.ValueEquals("Project"))
+                            {
+                                if (profile.Value.TryGetProperty("applicationUrl", out var applicationUrl))
+                                {
+                                    var value = applicationUrl.GetString();
+                                    if (value is { } applicationUrls)
+                                    {
+                                        var urls = applicationUrls.Split(";");
+                                        var firstHttpsUrl = urls.FirstOrDefault(u => u.StartsWith("https:", StringComparison.OrdinalIgnoreCase));
```

Comment:
Now that we're revisiting the HTTPS-defaults I'm wondering if we should relax this slightly by just having it use the first URL it finds. We actually might want to consider having it default to adding multiple audiences, i.e. one for each address it finds, so that the JWT will work with any configured launch profile of the target project.

Reply:
We can do this!

I discovered a bothersome gap in the JwtBearer API wherein you cannot pass multiple audiences into the creators/constructors for a JwtSecurityToken and instead have to register them as multiple `aud` claims. We have this now though:

```
"tre": {
        "Audiences": [
          "https://localhost:7157",
          "http://localhost:5157"
        ],
        "ClaimsIssuer": "dotnet-user-jwts"
      }
```